Classification demo (#639)
opencv/open_model_zoo: demos/imagenet_classification_demo/README.md

> # Imagenet Classification C++ Demo

Reviewer: The focus on ImageNet here seems unwarranted. Is the demo actually tied to ImageNet in any way? It seems that it should work with any network, as long as an appropriate file with labels is supplied.

# Imagenet Classification C++ Demo

The demo shows an example of using neural networks for image classification.

You can use the following pre-trained models with the demo:

* `resnet-50`
* `vgg19`
* `googlenet-v4`
* all other classification models (please, check [models.lst](./models.lst) file)

For more information about the pre-trained models, refer to the [model documentation](../../models/public/index.md).

## How It Works

> On the start-up, the application reads command line parameters and loads one network to the Inference Engine for execution. 

Reviewer 2: ```suggestion
On the start-up, the application reads command line parameters and loads a classification network to the Inference Engine for execution.
```

    -b "<integer>"            Optional. Specify batch to infer. Default value is 1.
    -nthreads "<integer>"     Optional. Specify count of threads.
    -nstreams "<integer>"     Optional. Specify count of streams.
    -nireq "<integer>"        Optional. Number of infer requests.
    -nt "<integer>"           Optional. Number of top results. Default value is 5. Must be >= 1.
    -res "<WxH>"              Optional. Set image grid resolution in format WxH. Default value is 1920x1080.
    -no_show                  Optional. Disable showing of processed images.
    -time "<integer>"         Optional. Time in seconds to execute program. Default is -1 (infinite time).
    -u                        Optional. List of monitors to show initially.
```

The number of InferRequests is specified by -nireq flag.
Each InferRequest acts as a "buffer": it waits in queue before being filled with images and sent for inference, then after
the inference completes, it waits in queue before its results would be processed.
Increasing the number of InferRequests usually increases performance, because in that case multiple InferRequests can be

> processed simultaneously if the device supports parallelization.

Reviewer 2: This way of line breaking looks strange. Looks that GitHub's preview width is slightly less than 120 symbols.

# Imagenet Classification C++ Demo

The demo shows an example of using neural networks for image classification.

You can use the following pre-trained models with the demo:

* `resnet-50`
* `vgg19`
* `googlenet-v4`

> * all other classification models (please, check models.lst file)

Reviewer 2: ```suggestion
* all other classification models (please, check [models.lst](./models.lst) file)
```

For more information about the pre-trained models, refer to the [model documentation](../../models/public/index.md).

## How It Works

On the start-up, the application reads command line parameters and loads one network to the Inference Engine for execution. Upon getting an image, it performs inference of classification and places the processed image on grid.

The demo starts in "Testing mode" with fixed grid size. After calculating the average FPS result, it will switch to
normal mode and grid will be readjusted depending on model performance. Bigger grid means higher performance.

The text above each image shows whether the classification was correct: green means right class prediction, red means wrong.

> Use "C", "D" and "M" keys to toggle resource motitors.

Reviewer 2: Adding other monitors will require updating that list. I suggest just avoiding introducing the list.

> You can stop the demo by pressing "Esc" button. After that, the average telemetry values would be printed to the console.

Reviewer 2: ```suggestion
You can stop the demo by pressing "Esc" button. After that, the average metrics values would be printed to the console.
```
metric is a more common term.

> **NOTE**: By default, Open Model Zoo demos expect input with BGR channels order. If you trained your model to work with RGB order, you need to manually rearrange the default channels order in the demo application or reconvert your model using the Model Optimizer tool with `--reverse_input_channels` argument specified. For more information about the argument, refer to **When to Reverse Input Channels** section of [Converting a Model Using General Conversion Parameters](https://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_convert_model_Converting_Model_General.html).

## Running

Running the application with the <code>-h</code> option yields the following usage message:
```sh
./imagenet_classification_demo -h

imagenet_classification_demo [OPTION]
Options:

    -h                        Print a usage message.

>     -i "<path>"               Required. Path to a folder with images or path to an image files: a .ubyte file for LeNet and a .bmp file for the other networks.

Reviewer 2: There is no LeNet in models.lst

    -b "<integer>"            Optional. Specify batch to infer. Default value is 1.
    -nthreads "<integer>"     Optional. Specify count of threads.
    -nstreams "<integer>"     Optional. Specify count of streams.
    -nireq "<integer>"        Optional. Number of infer requests.
    -nt "<integer>"           Optional. Number of top results. Default value is 5. Must be >= 1.
    -res "<WxH>"              Optional. Set image grid resolution in format WxH. Default value is 1920x1080.
    -no_show                  Optional. Disable showing of processed images.
    -time "<integer>"         Optional. Time in seconds to execute program. Default is -1 (infinite time).
    -u                        Optional. List of monitors to show initially.
```

The number of infer requests can be set by "-nireq" flag.
* small nireq: low FPS (device cores aren't fully loaded) and low latency (images don't wait for long before being sent
for inference).
* nireq = number of cores: high FPS (full device utilization), median latency (images wait for some time). Considered optimal.

> * nireq > number of cores: high FPS (same as in the previous case), high latency (images wait for long time).

Reviewer 2: The description shouldn't be based on a number of cores because it is harder to apply that term for some types of devices.
I would write about the idea of using InferRequests as buffers, so the larger -nireq the bigger the buffer and bigger the latency, but it becomes easier to parallelize the jobs.

As for gaining FPS, this is not about nireq compared to cores, nireq should slightly exceed nstreams value summed over all used devices. (And in case of CPU nstreams should be equeal to number of cores.)

I also doubt about `nireq = number of cores: high FPS (full device utilization)` assuming CPU only is used. The pipeline needs an additional element to pre/post process while other images are inferred, so setting that exact number is sub-optimal.

I would also add a note somewhere about the demo reading all the images before an actual execution, which may cause the demo to seem doing nothing for a user.

Running the application with the empty list of options yields an error message.

To run the demo, you can use public or pre-trained models. To download the pre-trained models, use the OpenVINO [Model Downloader](../../tools/downloader/README.md) or go to [https://download.01.org/opencv/](https://download.01.org/opencv/).

> **NOTE**: Before running the demo with a trained model, make sure the model is converted to the Inference Engine format (\*.xml + \*.bin) using the [Model Optimizer tool](https://docs.openvinotoolkit.org/latest/_docs_MO_DG_Deep_Learning_Model_Optimizer_DevGuide.html).

For example, use the following command line command to run the application:
```sh
./imagenet_classification_demo -m <path_to_classification_model> \
                      -i <path_to_folder_with_images> \
                      -labels <path_to_file_with_list_of_labels> \
                      -gt <path_to_file_with_list_of_classes> \
                      -u CDM
```

## Demo Output

The demo uses OpenCV to display the resulting image grid with classification results presented as a text above images.

> After the completion, it prints average telemetry values to the console.

Reviewer 2: ```suggestion
After the completion, it prints average metrics values to the console.
```

## Required files

If you want to see classification results, you must use "-gt" and "-labels" flags to specify two .txt files
containing lists of classes and labels.

> "Classes" file is used for matching image file names with right object classes.

Reviewer 2: ```suggestion
"Ground truth" file is used for matching image file names with correct object classes.
```

It has the following format:

./ILSVRC2012_val_00000001.JPEG 65

./ILSVRC2012_val_00000002.JPEG 970

./ILSVRC2012_val_00000003.JPEG 230

...

> "Labels" file contains the list of human-readable labels, one line for each class.

Reviewer 2: ```suggestion
["Labels" file](./synset_words.txt) contains the list of human-readable labels, one line for each class.
```

> You can find this file in source code: /open_model_zoo/demos/imagenet_classification_demos/synset_words.txt

Reviewer 2: ```suggestion
```

It has the following format:

n01440764 tench, Tinca tinca

n01443537 goldfish, Carassius auratus

> n01484850 great white shark, white shark, man-eater, man-eating shark, Carcharodon carcharias

Reviewer 2: So why was that format chosen?

Author: @Wovchena 
There are line-breaks just for better look with Markdown formatting of this README file.

The actual file is like this:
n01440764 tench, Tinca tinca
n01443537 goldfish, Carassius auratus
n01484850 great white shark, white shark, man-eater, man-eating shark, Carcharodon carcharias

Katya told me that this format is the most common, as far as she knows.
It's also possible to use this format:
n12345678 1 word1, word2
n12345678 2 word3, word4
but these additional numbers don't contain any useful information, so I think they're not needed.

Reviewer 2: Ok.
(The question was about including or not synset words.)

...

> **NOTE**: On VPU devices (Intel® Movidius™ Neural Compute Stick, Intel® Neural Compute Stick 2, and Intel® Vision Accelerator Design with Intel® Movidius™ VPUs) this demo has been tested on the following Model Downloader available topologies: 
>* `resnet-50`
>* `vgg19`
>* `googlenet-v4`
>

> > Other models may produce unexpected results on these devices.

Reviewer 2: I think we should drop it.
